crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,341 @@
+    fn add_ops(
+        &mut self,
+        op_list: Vec<OpId>,
+        source: AgentId,
+    ) -> BoxFut<'_, K2Result<()>> {
+        Box::pin(async move {
+            // Add ops to queue.
+            let mut ops = self.0.ops.lock().await;
+            op_list.clone().into_iter().for_each(|op_id| {
+                ops.insert((op_id, source.clone()), ());
+            });
+            drop(ops);
+
+            // Wake up fetch task in case it is sleeping. Time out if the channel buffer
+            // is full.
+            let _ = tokio::time::timeout(
+                Duration::from_millis(5),
+                self.0.fetch_request_tx.send(()),
+            )
+            .await;
+
+            Ok(())
+        })
+    }
+}
+
+#[derive(Clone, Debug)]
+struct Inner {
+    config: CoreFetchConfig,
+    // An `IndexMap` is used to retain order of insertion and have the ability to look up elements
+    // by key efficiently. Ops may be added redundantly to the map with different sources to fetch from, so
+    // the map is keyed by op and agent id together.
+    // The value field could be used to track number of attempts for example.
+    ops: Arc<Mutex<IndexMap<(OpId, AgentId), ()>>>,
+    current_request_count: Arc<Mutex<u8>>,
+    cool_down_list: Arc<Mutex<HashMap<AgentId, Instant>>>,
+    // A sender to wake up a sleeping fetch task.
+    fetch_request_tx: Sender<()>,
+}
+
+impl Inner {
+    pub fn spawn_fetch_task(
+        config: CoreFetchConfig,
+        transport: DynTransport,
+    ) -> Self {
+        // Create a channel to wake up sleeping loop.
+        let (fetch_request_tx, mut fetch_request_rx) =
+            tokio::sync::mpsc::channel::<()>(50);
+
+        let inner = Self {
+            config,
+            ops: Arc::new(Mutex::new(IndexMap::new())),
+            current_request_count: Arc::new(Mutex::new(0)),
+            cool_down_list: Arc::new(Mutex::new(HashMap::new())),
+            fetch_request_tx,
+        };
+
+        tokio::spawn({
+            let current_request_count = inner.current_request_count.clone();
+            let inner = inner.clone();
+            let transport = transport.clone();
+
+            async move {
+                let CoreFetchConfig {
+                    parallel_request_count,
+                    parallel_request_pause,
+                    fetch_loop_sleep,
+                    ..
+                } = inner.config;
+
+                // Loop for processing op/agent pairs in the queue. Runs continuously and is paused
+                // briefly when the maximum of parallel requests is reached, and sleeps for a
+                // longer period when the queue is empty. It is awoken from sleep through a channel
+                // when new op ids are added to the queue.
+                loop {
+                    // Send new request if parallel request count is not reached.
+                    if parallel_request_count
+                        > (*current_request_count.lock().await)
```

Comment:
I find that it's easier to understand parallelism when there are actually parallel tasks: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=606abba6cb763c472a2037e6bf24e1ce

This way when we get to awaiting the response, it can just do that inline with a timeout and you won't need your waker channel anymore. If you get a timeout, mark the agent unreliable and send the op into limbo. Otherwise pull the next item to fetch.

Reply:
Yes, that seems like the missing piece I had in my thinking. That looks much easier to understand. There are 2 tasks which are the 2 parallel requesters, I understand. In the actual code the `for 0..5` loop is replaced by just a `loop {}`, right?

Reply:
Yeah, actually probably a `while let Some(bla) = ` loop. The for was just so the example would end : )